[sharded_tensor] add new init_from_local_shards API

wanchaol opened the pull request:
Summary: This added `init_from_local_shards` API to construc t a ShardedTensor from local_shards and global metadata list.

Test Plan: 
test_init_from_local_shards
test_init_from_local_shards_invalid_sharding

Differential Revision: D29276777

wanchaol commented on torch/distributed/_sharded_tensor/api.py:
```diff
@@ -214,6 +223,76 @@ def _init_rpc(self):
+    def init_from_local_shards(
+        cls,
+        local_shards: List[torch.Tensor],
+        shards_metadata: List[ShardMetadata],
```
@pritamdamania I am thinking of whether we should let user pass in `EnumerableShardingSpec` here instead of `List[ShardMetadata]` as they are essentially the same thing. Let me know which one you prefer.

wanchaol replied:
Yeah I think adding a `ShardedTensorMetadata` make a lot of sense to me, and grouping the attributes together makes it more intuitive to the user as well I think. This might makes it looks similar but not entirely same concept with the ShardingSpec, but I think we can document it so that user won't get confused.

wanchaol commented on torch/distributed/_sharded_tensor/api.py:
```diff
@@ -214,6 +223,76 @@ def _init_rpc(self):
+    ):
+
+        if len(local_shards) == 0 or len(shards_metadata) == 0:
+            raise ValueError("local_shards or shards_metadata must not be empty!")
```
@pritamdamania I added the check here to check if `local_shards` is empty for each rank to call this. Do you think this is a valid check, or are there cases that in some rank `local_shards` can be empty? (i.e. shard_metadata.shard_lengths = [0, 0, ..] for the shard metadata in that rank)

pritamdamania87 commented on torch/distributed/_sharded_tensor/api.py:
```diff
@@ -214,6 +223,76 @@ def _init_rpc(self):
+            # validate local_shards and metadata
+            if current_rank == rank:
+                local_shard = local_shards[local_shard_idx]
+                if shard_metadata.shard_lengths != list(local_shard.size()):
```
Isn't this assuming some sort of order between shard_metadata and local_shards? For example doesn't this fail if the first shard_metadata we encounter is for local_shards[1]? Local shards could have different shard_lengths.

wanchaol replied:
yep here we are assuming the order passed in between `local_shards` and `shards_metadata` are consistent, this is a requirement from the user, which might be not that nice, but we have to specify the order in order for the corresponding metadata and local shard to match, I couldn't think of a better idea rather than pushing this requirement to the user yet, do you have any insights on how we could match them seamlessly without imposing the order?

pritamdamania87 replied:
The shard_metadata contains the location of all shards, based on that you can figure out which metadata belongs to this rank and then validate that with local_shards?

wanchaol replied:
Yeah we can do that, the ordering of metadata won't matter across locations as we can figure the location out from shard_metadata and assert the local tensor is on the same location. But the problem still happens if we have multiple tensors in the same location, consider only one location `cuda:0`, suppose there're two shards/tensors for location `cuda:0`, there're two shard_metadatas associated with them, and if they are the same size, how do we know which tensor match to which shard_metadata? The current approach I am doing for this case is, since the shard_metadata passed in is a list, the first occuring shard_metadata that has location `cuda:0` should match to the first tensor,

pritamdamania87 commented on torch/distributed/_sharded_tensor/api.py:
```diff
@@ -214,6 +223,76 @@ def _init_rpc(self):
+        num_dims = len(shards_metadata[0].shard_lengths)
+        offset_length_tuple_list_by_dim: List[List[Tuple[int, int]]] = [[] for _ in range(num_dims)]
+        # validate the metadata and add to local_shards
+        for shard_metadata in shards_metadata:  # type: ignore[attr-defined]
```
Shouldn't we do some validation on shards_metadata? For example, what if the user specifies a remote device that does not exist?

pritamdamania87 commented on torch/distributed/_sharded_tensor/api.py:
```diff
@@ -214,6 +223,76 @@ def _init_rpc(self):
+
+        # recover the global dims information
+        dims = [0] * num_dims
+        for i, offset_length_tuple_list in enumerate(offset_length_tuple_list_by_dim):
```
Don't we need to do somme validation that there are no "gaps" between the shards? I'm wondering if we should add `ShardedTensorMetadata` as part of this change itself since it would be much more clear in terms of validation.

For example, ShardedTensorMetadata would contain the overall_dims of the tensor and that can be easily verified with shards_metadata.

wanchaol replied:
yeah I think this make sense, if we have ShardedTensorMetadata, we only need to verify the dims constructed by List[ShardedMetadata] matches overall_dims. In terms of "gaps" validation, we do have it checked here and raise error when `dims[i] < offset`

pritamdamania87 commented on torch/distributed/_sharded_tensor/api.py:
```diff
@@ -214,6 +223,76 @@ def _init_rpc(self):
+            for (offset, length) in offset_length_tuple_list:
+                if dims[i] == offset:
+                    dims[i] += length
+                elif dims[i] > offset:
```
Do we handle overlapping shards? In terms of validation see https://github.com/pytorch/pytorch/blob/master/torch/distributed/_sharding_spec/api.py#L208, we can probably reuse some of that code.

pritamdamania87 commented on test/distributed/_sharded_tensor/test_sharded_tensor.py:
```diff
@@ -799,3 +799,107 @@ def test_with_rpc_names(self):
+    @with_comms
+    @skip_if_lt_x_gpu(4)
+    @requires_nccl()
+    def test_init_from_local_shards(self):
```
Can we add the following tests:

1. Some ranks don't have local shards.
2. Use a process group which is subset of default group using `new_group`.
3.  overlapping shards.
4. shards with gaps.

pritamdamania87 commented on torch/distributed/_sharded_tensor/api.py:
```diff
@@ -237,6 +253,83 @@ def _init_rpc(self):
+                    f'local ShardMetadata shard lengths: {shard_meta.shard_lengths}'
+                )
+
+            if str(local_shard_tensor.device) != local_device:
```
We should also verify that the `dtype`, `layout`, `requires_grad` etc. of local_shards is same as what is specified in ShardedTensorMetadata.

Can we also add unit tests where this contract is not respected and we throw an appropriate error?

wanchaol replied:
good point, just added

pritamdamania87 commented on torch/distributed/_sharded_tensor/__init__.py:
```diff
@@ -48,3 +48,33 @@ def empty(
+        sharded_tensor_metadata (:class:`torch.distributed._sharded_tensor.ShardedTensorMetadata`)
+            The ShardedTensorMetadata that created manually, represents the global metadata
+            of the ShardedTensor, must comply with `local_shards` defined in each rank.
+            Note that `sharded_tensor_metadata` must be valid and should also contains
```
nit: s/contains/contain/g

pritamdamania87 reviewed (commented):
Looks great overall, just had one comment inline that needs to be addressed.

pritamdamania87 commented on torch/distributed/_sharded_tensor/api.py:
```diff
@@ -246,6 +262,104 @@ def _init_rpc(self):
+                    f'sharded_tensor_metadata layout: {sharded_tensor_metadata.layout}'
+                )
+
+            if local_shard_tensor.requires_grad != sharded_tensor_metadata.requires_grad:
```
Can't we verify memory layout as well? Since we only support `contiguous`, shouldn't `torch.is_contiguous` be sufficient?

pritamdamania87 commented on torch/distributed/_sharded_tensor/api.py:
```diff
@@ -246,6 +262,104 @@ def _init_rpc(self):
+                )
+
+            if local_shard_tensor.requires_grad != sharded_tensor_metadata.requires_grad:
+                raise ValueError(
```
Can't we also verify pin_memory using `tensor.is_pinned`?

Commit:
[sharded_tensor] add new init_from_local_shards API (#60479)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/60479

This added `init_from_local_shards` API to construct a ShardedTensor from local_shards and global sharded_tensor_metadata. It also refactors the utils in ShardingSpec to be able to be used by sharded_tensor for sanity check purpose.

Test Plan:
test_init_from_local_shards
test_init_from_local_shards_invalid_sharding

Reviewed By: pritamdamania87

Differential Revision: D29276777

fbshipit-source-id: a6f709a03ca54ab119a40d3abd90f3c80bcc26ab
